### PR TITLE
Task attachments added, test adjusted. 

### PR DESCRIPTION
**Pull request by aliilyas**
Solution for issue #32

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -9,4 +9,6 @@ services:
       - POSTGRES_USER=tiny_task
       - POSTGRES_PASSWORD=demo123
     volumes:
-      - ./postgres/data:/var/lib/postgresql/data
+      - pgdata:/var/lib/postgresql/data
```
Comment (GeBeater) [inline docker-compose.yml]: not part of task #32. what is the reason for this change and why volumes are used over mounts?

File: build.gradle [hunk 2]
```diff
@@ -25,6 +25,7 @@ dependencies {
   implementation 'org.flywaydb:flyway-core'
   implementation 'net.rakugakibox.spring.boot:orika-spring-boot-starter:1.8.0'
   implementation 'org.postgresql:postgresql'
```
Comment (GeBeater) [inline build.gradle]: commons-io:commons-io or rather org.apache.commons.io.FilenameUtils would no longer be required by applying the commentary listed below

File: src/main/java/com/coyoapp/tinytask/dto/TaskResponse.java [hunk 3]
```diff
@@ -14,4 +14,6 @@
   private String id;
 
   private String name;
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/dto/TaskResponse.java]: due to the adjustments in the DefaultFileService it is more a unique file identifier instead of the origin filename

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -3,12 +3,26 @@ spring:
...
       ddl-auto: validate
+    properties:
+      hibernate:
+        temp:
+          use_jdbc_metadata_defaults: false
 
 logging:
   level:
     com.coyoapp.tinytask: DEBUG
+
```
Comment (GeBeater) [inline src/main/resources/application.yml]: backend does not start out of the box due to /storage/.. is read-only. fixed it by create directory /tmp/tiny-task on local machine. an adjustment of the README that the path must be readable or changed would be great

File: src/main/java/com/coyoapp/tinytask/repository/DefaultFileRepository.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+
+@Component
+public class DefaultFileRepository implements FileRepository {
+
+  @Value("${file.base.dir}")
+  private String filesBaseDir;
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/repository/DefaultFileRepository.java]: I would simplify and rename the method like below.

```
  public Path createFileBaseDirectoryIfNotExists() {
    try {
      return Files.createDirectories(Paths.get(filesBaseDir));
    } catch (IOException e) {
      throw new IllegalArgumentException("Could not create file base dir", e);
    }
  }
```

this optional makes the code unnecessarily complicated. especially an exception is thrown anyway if the directory could not be created (see DefaultFileService.initImageRootFolder())

File: src/main/java/com/coyoapp/tinytask/repository/DefaultFileRepository.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    if (!Files.exists(path)) {
+      try {
+        path = Files.createDirectories(path);
+      } catch (IOException e) {
+        return Optional.empty();
+      }
+    }
+    return Optional.ofNullable(path);
+  }
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/repository/DefaultFileRepository.java]: same argumentation regarding optional here. I would simplify the code like below.

```
  public Path saveFile(String name, byte[] content) {
    try {
      Path destination = Files.createFile(Paths.get(filesBaseDir, name));
      try (FileOutputStream fos = new FileOutputStream(destination.toFile())) {
        fos.write(content);
        return destination;
      }
    } catch (Exception e) {
      throw new FileException("Could not create file");
    }
  }
```

File: src/main/java/com/coyoapp/tinytask/repository/DefaultFileRepository.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      Path destination = Files.createFile(Paths.get(filesBaseDir, name));
+      try (FileOutputStream fos = new FileOutputStream(destination.toFile())) {
+        fos.write(content);
+      }
+      return Optional.ofNullable(destination);
+    } catch (Exception e) {
+      return Optional.empty();
+    }
+  }
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/repository/DefaultFileRepository.java]: same argumentation for optionals here. would simplify the code (no example here; inspire by examples above).

File: src/main/java/com/coyoapp/tinytask/repository/FileRepository.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package com.coyoapp.tinytask.repository;
+
+import java.nio.file.Path;
+import java.util.Optional;
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/repository/FileRepository.java]: the repositories in this package are spring data repository. could be confuse for someone.

File: src/main/java/com/coyoapp/tinytask/repository/FileRepository.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+public interface FileRepository {
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/repository/FileRepository.java]: optionals can removed, see comments above

File: src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.Optional;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class DefaultFileService implements FileService {
+
+  @Autowired
+  private FileRepository fileRepository;
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java]: due to comment above, method can be simplified by ...

```
  @PostConstruct
  private void initImageRootFolder() {
    fileRepository.createFileBaseDirectoryIfNotExists();
  }
```

File: src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+  @PostConstruct
+  private void initImageRootFolder() {
+    Path path = fileRepository.initFilesDir()
+      .orElseThrow(() -> new RuntimeException("File base directory not accesssible."));
+    log.debug("initImageRootFolder(imageBaseDir={})", path.toAbsolutePath());
+  }
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java]: due to comment above, method can be simplified by ...

```
  public String saveFile(MultipartFile file) {
    String fileName = UUID.randomUUID().toString();
    try {
      fileRepository.saveFile(fileName, file.getBytes());
    } catch (Exception e) {
      throw new FileException("File saving failed.");
    }
    return fileName;
  }
```

File: src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        MultipartFile multipartFile = file.get();
+        fileName = wrapFileName(multipartFile);
+        fileRepository.saveFile(fileName, multipartFile.getBytes()).orElseThrow(() -> new FileException("File saving failed."));
+      } catch (IOException e) {
+        throw new FileException("File reading failed.");
+      }
+    }
+    return fileName;
+  }
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java]: due to comment above, method can be simplified by ...

```
  public byte[] getFile(String fileName) {
    return fileRepository.getImageByName(fileName);
  }
```

File: src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  public byte[] getFile(String fileName) {
+    return fileRepository.getImageByName(fileName).orElseThrow(() -> new FileException("File reading error."));
+  }
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java]: this looks like, that the filename was used to generate a unique file identifier. thus I would simplify the code by using UUID.randomUUID().toString() instead

File: src/main/java/com/coyoapp/tinytask/service/DefaultTaskService.java [hunk 14]
```diff
@@ -20,13 +23,16 @@
 public class DefaultTaskService implements TaskService {
 
   private final TaskRepository taskRepository;
+  private final FileService fileService;
   private final MapperFacade mapperFacade;
 
   @Override
   @Transactional
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/service/DefaultTaskService.java]: optionals as parameter is generally not the best idea. i ask you to do some research on this yourself, so that I can skip the argumentation here.

File: src/main/java/com/coyoapp/tinytask/service/TaskService.java [hunk 15]
```diff
@@ -2,11 +2,14 @@
 
 import com.coyoapp.tinytask.dto.TaskRequest;
 import com.coyoapp.tinytask.dto.TaskResponse;
+import org.springframework.web.multipart.MultipartFile;
+
 import java.util.List;
+import java.util.Optional;
 
 public interface TaskService {
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/service/TaskService.java]: see "optionals as parameter" comment above

File: src/main/java/com/coyoapp/tinytask/web/TaskController.java [hunk 16]
```diff
@@ -27,9 +24,9 @@
   private final TaskService taskService;
 
   @PostMapping
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/web/TaskController.java]: javax.validation.constraints for @Valid no longer works due to usage of @RequestParam

File: src/main/java/com/coyoapp/tinytask/repository/DefaultFileRepository.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
+package com.coyoapp.tinytask.repository;
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/repository/DefaultFileRepository.java]: no tests for DefaultFileRepository available. (does not reflect skill sheet rating (especially for TTD) 😉)

File: src/test/java/com/coyoapp/tinytask/service/DefaultFileServiceTest.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+import org.mockito.junit.MockitoRule;
+import org.mockito.quality.Strictness;
+
+import java.util.Optional;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.when;
+
```
Comment (GeBeater) [inline src/test/java/com/coyoapp/tinytask/service/DefaultFileServiceTest.java]: only getFile of DefaultFileService is tested. test coverage does not reflect skill sheet rating (especially for TTD) 😉

File: src/test/java/com/coyoapp/tinytask/service/DefaultFileServiceTest.java [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+  @InjectMocks
+  private DefaultFileService objectUnderTest;
+
+  @Rule
+  public TemporaryFolder baseFileDir = new TemporaryFolder();
+
+  @Test
+  public void shouldReadFromRepo() {
+    //when
```
Comment (GeBeater) [inline src/test/java/com/coyoapp/tinytask/service/DefaultFileServiceTest.java]: filename not considered for test

File: src/test/java/com/coyoapp/tinytask/service/DefaultFileServiceTest.java [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+import org.mockito.quality.Strictness;
+
+import java.util.Optional;
+
```
Comment (GeBeater) [inline src/test/java/com/coyoapp/tinytask/service/DefaultFileServiceTest.java]: hamcrest assertThat and mockito matchers instead of AssertJ (see already existing tests)

File: src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+import java.io.IOException;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.Optional;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class DefaultFileService implements FileService {
+
```
Comment (GeBeater) [inline src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java]: constructor injection should be preferred (see other components) instead of field injection

File: src/test/java/com/coyoapp/tinytask/web/TaskControllerTest.java [hunk 22]
```diff
@@ -50,6 +45,29 @@ public void shouldCreateTask() throws Exception {
       .andExpect(jsonPath("$.name", is(name)));
   }
 
+  @Test
```
Comment (GeBeater) [inline src/test/java/com/coyoapp/tinytask/web/TaskControllerTest.java]: I would prefer to extend the existing shouldCreateTask test instead of adding the shouldCreateTaskWithFile (for this testing purpose)

File: src/main/webapp/app/app.module.ts [hunk 23]
```diff
@@ -4,9 +4,10 @@ import { MatButtonModule } from '@angular/material/button';
 import { MatIconModule } from '@angular/material/icon';
```
Comment (GeBeater) [inline src/main/webapp/app/app.module.ts]: I skipped the frontend review remarks in general. unfortunately there are also some things to note

File: src/main/resources/application.yml (hunk 4 above)
Comment (aliilyas) [reply src/main/resources/application.yml]: Did not pay attention to that, tried to make less config dependent solution.

File: src/main/java/com/coyoapp/tinytask/web/TaskController.java (hunk 16 above)
Comment (aliilyas) [reply src/main/java/com/coyoapp/tinytask/web/TaskController.java]: Roger that! missed that.

File: src/main/java/com/coyoapp/tinytask/service/DefaultTaskService.java (hunk 14 above)
Comment (aliilyas) [reply src/main/java/com/coyoapp/tinytask/service/DefaultTaskService.java]: It depends on project coding conventions. There is no blocker argument not to use that.

File: src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java (hunk 13 above)
Comment (aliilyas) [reply src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java]: Nope, UUIDs instead of readable file names are harder to read for users. Just pretend you downloaded attachments from a couple of tasks and trying to identify them in the downloads folder. (Sorry that's my nightmare XD )

File: src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java (hunk 11 above)
Comment (aliilyas) [reply src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java]: "File reading failed." is wrong should be "File reading input MultipartFile failed.", `multipartFile.getBytes()` can throw IOException. 
`fileRepository.saveFile` can throw a bulk of different exceptions, which are simplified to single optional to make raise FileException fluent one-liner.

File: src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java (hunk 21 above)
Comment (aliilyas) [reply src/main/java/com/coyoapp/tinytask/service/DefaultFileService.java]: Roger that!

File: src/main/java/com/coyoapp/tinytask/repository/FileRepository.java (hunk 8 above)
Comment (aliilyas) [reply src/main/java/com/coyoapp/tinytask/repository/FileRepository.java]: Creating another repository package will be more confusing. By default repository is a data access layer, in this case, implementation of a repository can be changed to use database storage, sending file to some cloud storage.
